### PR TITLE
AST: Fix bug in subclass composition minimization

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
Now that opened archetype signatures are generalized to encompass outer generic contexts, the number of requirements in the generic signature is not always indicative of whether the minimal type is also a composition.

### CODE REVIEW

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

File: lib/AST/Type.cpp [hunk 1]
```diff
@@ -4126,10 +4125,6 @@ CanType ProtocolCompositionType::getMinimalCanonicalType(
 
     switch (Req.getKind()) {
     case RequirementKind::Superclass:
-      assert((!superclass || superclass->isEqual(Req.getSecondType()))
-             && "Multiple distinct superclass constraints!");
-      superclass = Req.getSecondType();
```
Comment (slavapestov) [inline lib/AST/Type.cpp:4131]: This is a nice simplification but I don't understand why it was failing before. We already skip requirements where the subject type is not equal to the last generic parameter. There should only ever be at most one superclass requirement that reaches this point.

Comment (AnthonyLatsis) [reply lib/AST/Type.cpp:4131]: The simplification is NFC. 

> This is a nice simplification but I don't understand why it was failing before. 

If the generic signature has more than one requirement, and only one relevant requirement on the last generic parameter, the `Reqs.size() == 1` check does not catch the single-member composition and we hit the corresponding assertion in `ProtocolCompositionType::build`:

```swift
protocol P4 {}
class C: P4 {}
struct S<T: P4> {
  func foo(_: C & P4) {} // Minimal type is C
}
```

**Review by slavapestov — APPROVED** [review]

File: lib/AST/Type.cpp [hunk 2]
```diff
@@ -4141,10 +4136,15 @@ CanType ProtocolCompositionType::getMinimalCanonicalType(
...
-  // the signature's requirements.
-  if (superclass)
-    MinimalMembers.insert(MinimalMembers.begin(), superclass->getCanonicalType());
+  // A superclass constraint is always retained and must appear first in the
+  // members list.
+  assert(Composition->getMembers().front()->getClassOrBoundGenericClass() ==
+         MinimalMembers.front()->getClassOrBoundGenericClass());
+
+  // If we are left with a single member and no layout constraint, the member
+  // is the minimal type.
```
Comment (slavapestov) [inline lib/AST/Type.cpp:4145]: Maybe add a comment here that this check is needed since build() cannot be called with a single member.

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Pull request merged**